### ISSUE TITLE
Enable offline support for all DevTools screens

### ISSUE THREAD

**Issue by kenzieschmoll** [flutter/devtools#4470]
Enabling offline support will:

1. Be helpful for users who want to export data and reload later or to send to other developers for debugging
2. Be _especially_ helpful for faking test data while developing DevTools. Then we can just write `SomeScreenData.parse(someOfflineData)` to initialize fake testing data for the screen. If we continue to integrate with Stager, this will be useful for both testing and stager development (see [https://github.com/flutter/devtools/pull/446](https://github.com/flutter/devtools/pull/4463))
3. Be helpful for users to be able to include in bug reports

- [ ] Inspector - https://github.com/flutter/devtools/issues/1966
- [ ] Performance - already done for the most part, but need to make sure we include any raster metrics data that is present 
- [ ] CPU Profiler - already done, but need to ensure we are including any filters that are set on the data
- [ ] Memory Profiler - need to switch to how other screens export data and send data to Downloads (https://github.com/flutter/devtools/issues/3411) and need to audit to ensure that all data from screen is included in the output
- [ ] Debugger - is there a use case for looking at an "offline" debugger state? This is still useful for creating test data files, but we may want to make these methods visibleForTesting and potentially only expose for debug builds of the DevTools app if we don't want to expose an "Export" button for users.
- [ ] Network - we have an "Export as Har" function, but we need to hook this up to our existing import / export framework and ensure that all data to recreate the state of the screen for offline uses is included in the output (e.g. selected request, filters, etc.)
- [ ] Logging
- [ ] App Size
- [ ] Vm Developer Tools Screen

**Comment by CoderDake** [comment]
It was mentioned that the Logging page might be a good place to start.

**Comment by CoderDake** [comment]
For the Network Page, consider making sure that we use HAR file format for network requests

**Comment by kenzieschmoll** [comment]
Exporting as HAR already exists for the network page, but for the full export of the network page, we'll also need to include extra devtools-specific information (things like selected request, filters, etc.).

**Comment by polina-c** [comment]
Related issue: https://github.com/flutter/devtools/issues/4645

**Comment by hrajwade96** [comment]
@kenzieschmoll,  in reference to our earlier discussion (https://github.com/flutter/devtools/pull/7970#discussion_r1664777833), what is the plan for adding offline support to the network page? I am interested in contributing to this effort and would like to know how I can help.

**Comment by kenzieschmoll** [comment]
To implement offline support for a screen, you can follow the instructions in the dart doc for `OfflineScreenControllerMixin`: https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/shared/offline_data.dart/#L125

You can look at [ProfilerScreenController](https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/screens/profiler/profiler_screen_controller.dart/#L24) and [PerformanceController](https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/screens/performance/performance_controller.dart/#L33) as examples.

> And add file extension detection (do explain what this detection is needed for)

The more complicated part for the Network page is the fact that we support exporting as a .har file, whereas the offline framework for DevTools only supports importing & exporting to / from a JSON file. So in the current form of the offline framework, if network data is exported as a .har file, it will not be able to be re-imported back into DevTools. Only data exported as a JSON file through the DevTools offline framework can be re-imported back into DevTools.

We can tweak the logic to support loading the .har file for the network page as a special case, but we can do this as a follow up task to adding the full offline support for the Network page. The logic to import data from a file is here: https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/shared/config_specific/import_export/import_export.dart/#L61-L98

**Comment by hrajwade96** [comment]
> To implement offline support for a screen, you can follow the instructions in the dart doc for `OfflineScreenControllerMixin`: https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/shared/offline_data.dart/#L125
> 
> You can look at [ProfilerScreenController](https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/screens/profiler/profiler_screen_controller.dart/#L24) and [PerformanceController](https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/screens/performance/performance_controller.dart/#L33) as examples.
> 
> > And add file extension detection (do explain what this detection is needed for)
> 
> The more complicated part for the Network page is the fact that we support exporting as a .har file, whereas the offline framework for DevTools only supports importing & exporting to / from a JSON file. So in the current form of the offline framework, if network data is exported as a .har file, it will not be able to be re-imported back into DevTools. Only data exported as a JSON file through the DevTools offline framework can be re-imported back into DevTools.
> 
> We can tweak the logic to support loading the .har file for the network page as a special case, but we can do this as a follow up task to adding the full offline support for the Network page. The logic to import data from a file is here: https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/shared/config_specific/import_export/import_export.dart/#L61-L98

Thanks for sharing the details, 
I had noticed that at present only exported 'json' format is supported for importing, I was thinking to make it accept 'har' as well. But as you mentioned, I will raise a separate PR for that change.
I will work on adding offline support with below feature for network screen - when the app gets disconnected, view offline data and check the instructions and references shared by you.

**Comment by hrajwade96** [comment]
@kenzieschmoll any on going work or plans for remaining screens? What would be the next priority item from the remaining ones?

**Comment by kenzieschmoll** [comment]
Remaining priority:
1. Inspector. This one is high impact and we should already have some prior art here since we fetch the tree from the Flutter framework in a JSON format already. @elliette can provide more info on the current implementation. 
2. Logging. This is likely much simpler than the Inspector.
3. All others (we may decide that offline support is not necessary for these screens, so before starting work here, please start a discussion first.)

**Comment by hrajwade96** [comment]
@kenzieschmoll Thanks for outlining the priorities! I'd like to contribute once I have some bandwidth.
